scipy/stats/_stats.pyx

```diff
@@ -476,4 +479,42 @@ def _local_correlations(distx, disty, global_corr='mgc'):
     corr_mat[local_varx <= 0, :] = 0
     corr_mat[:, local_vary <= 0] = 0
 
-    return corr_mat
\ No newline at end of file
+    return corr_mat
+
+
+cpdef double geninvgauss_logpdf(double x, double p, double b):
+    return _geninvgauss_logpdf_kernel(x, p, b)
+
+
+cdef _geninvgauss_logpdf_kernel(double x, double p, double b):
+    cdef double z, c
+
+    z = cs.kve(p, b)
+    if math.isinf(z):
+        return NAN
+
+    if x <= 0:
+        return -INFINITY
+
+    c = -math.log(2) - math.log(z) + b
+    return c + (p - 1)*math.log(x) - b*(x + 1/x)/2
+
+
+cdef double _geninvgauss_pdf(double x, void *user_data) except *:
+    # destined to be used in a LowLevelCallable
+    # can't use nogil because cs.kve isn't marked nogil
```

Comment:
That function eventually calls `cython_special.kve` which isn't marked as nogil, so this function can't be marked as nogil:
```
Error compiling Cython file:
------------------------------------------------------------
...
        return 0.

    p = (<double *>user_data)[0]
    b = (<double *>user_data)[1]

    return math.exp(_geninvgauss_logpdf_kernel(x, p, b))
                                             ^
------------------------------------------------------------

_stats.pyx:514:46: Coercion from Python not allowed without the GIL
```

Reply:
Which is exactly my question: why is cs.kve not marked nogil? I thought all of cython-special is?

Reply:
I realize it's OT for this PR --- it's just that this PR is going to be setting a pattern, and it's nice to understand the details. 
Ditto for the pattern of vectorization, you seem to be vectorizing it with np.vectorize, is it intentional? Is it required? Is there a better way?

None of these is blocking this PR though

Reply:
Or 
```
------------------------------------------------------------
...
    cdef double z, c

    if x <= 0:
        return -INFINITY

    z = cs.kve(p, b)
             ^
------------------------------------------------------------

_stats.pyx:495:14: Coercion from Python not allowed without the GIL
```
I don't know enough cython at the moment to figure this one out.

Reply:
Perhaps someone with cython_special expertise could chip in

Reply:
It turns out that if one uses `cimport` one can mark it nogil. Just made that change.

> I realize it's OT for this PR --- it's just that this PR is going to be setting a pattern, and it's nice to understand the details.
Ditto for the pattern of vectorization, you seem to be vectorizing it with np.vectorize, is it intentional? Is it required? Is there a better way?

Not sure I wanted to set a trend by cythonising, I just wanted to relieve the pain point with the duration of the test suite.
All the PDFs/CDFs need to be vectorised over the arguments (x, p, b can all be arrays that can broadcast over each other), using the `vectorize` decorator is [used in the super class](https://github.com/scipy/scipy/blob/master/scipy/stats/_distn_infrastructure.py#L1605) to vectorise `cdf`. The `_continuous_distns.py` defines its own `_vectorize` decorator, but it's not clear to me what the advantage is to that, or why the straight-up numpy decorator isn't sufficient.

Reply:
> Which is exactly my question: why is cs.kve not marked nogil? I thought all of cython-special is?

I thought so too but maybe not? Looks like `kve` is, depending on the call, either of these:
https://github.com/scipy/scipy/blob/c1372d8aa90a73d8a52f135529293ff4edb98fc8/scipy/special/functions.json#L895-L896
and it doesn't look like any of the amos functions are `nogil`-ed, or at least I don't see an equivalent `.pxd` file like what cephes uses. I'm less familiar with the amos functions but those look like they are from fortran, not sure if that has anything to do with the nogil issue though.

Reply:
> It turns out that if one uses `cimport` one can mark it nogil. Just made that change.
> 
> > I realize it's OT for this PR --- it's just that this PR is going to be setting a pattern, and it's nice to understand the details.
> > Ditto for the pattern of vectorization, you seem to be vectorizing it with np.vectorize, is it intentional? Is it required? Is there a better way?
> 
> Not sure I wanted to set a trend by cythonising, I just wanted to relieve the pain point with the duration of the test suite.

Is the bottleneck in the python code here in the stats package or the lack of gil release in the `kve` function or both?
 
> All the PDFs/CDFs need to be vectorised over the arguments (x, p, b can all be arrays that can broadcast over each other), using the `vectorize` decorator is [used in the super class](https://github.com/scipy/scipy/blob/master/scipy/stats/_distn_infrastructure.py#L1605) to vectorise `cdf`. The `_continuous_distns.py` defines its own `_vectorize` decorator, but it's not clear to me what the advantage is to that, or why the straight-up numpy decorator isn't sufficient.

Hmm, not clear to me either, I like the speedup in the PR.  
As @ev-br says this could set a trend so my question is why not put the faster functions into `scipy.special` and call them here? 

That seems to be the pattern already followed by several of the distributions in `stats`, e.g. `incbet`, `bdtr`, `pdtr` etc.
I'm just thinking it gets kind of confusing to track if you've code cython in `stats` and you've got c/fortran/cython in `special` being called in `stats`. 

Hence my question, can we not put these nice, accelerated functions into `special` or is there a reason why not to?

Reply:
@andyfaff ok, the reasoning makes sense. Thanks for clarifying.